tikv/tikv
# Foreground Quota Limiter: limit the cpu, bandwidth to provide stable QPS

### What is changed and how it works?

Issue Number: Close #11855 #11856

What's Changed:


```commit-message
Add a global forefront quota limiter based speed limiter to record the CPU resource overhead occupied by various requests. When the limiter reaches the limit quota, the request will be forced to sleep for a period of time to compensate. Statistics are only done on the part of process and are therefore approximate.
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)

we can see after quota a small quota, QPS became more stable.
![image](https://user-images.githubusercontent.com/18193301/158135796-80771917-35f8-43b8-b17c-73239c51eeee.png)

We have some discussion on sleep using sync or async. If use sync mode, the related thread will be blocked but this may be what this feature want, which means compensating for the cost cpu time to yield cpu resource to other worker. If use async mode, await makes the threads not blocked but the CPU limit is inaccurate because `await` is just a task to give up the CPU, the worker still can execute others. Below is some comparative tests

1. Using smaller refills incurs more thread switching overhead
2. Check that the return value of each consume_write/read does not exceed the upper limit, this behavior has no benefits and no disadvantages, so I think it will help not to delay some requests for too long
3. In async delay mode, the quota_delay of `kv_snapshot` should not be delayed, because `kv_snapshot` is often called by `with_tls_engine`, which is part of other requests and is very frequent. In async mode, the limiter's buckets have not been refilled, so a large number of `kv_snapshots` are slowed down.
4. In sync sleep mode, cpu resources will be released truely, so the throttled duration on the grafana panel are more accurate and much smaller. Since async only kicks out the current tasks, it needs to artificially limit a smaller quota value to achieve a similar cpu limit effect to sync sleep mode.
5. Using speed limiter for cpu usage is not a good choice, because the time used by cpu itself also occupies time buckets, so maybe it should be delayed = (quota_delay - cost time)

![image](https://user-images.githubusercontent.com/18193301/158501214-2dc83c0d-1af4-492c-9e98-9b2a86e1762d.png)
![image](https://user-images.githubusercontent.com/18193301/158501326-19690c61-e907-40c7-aebe-095062ac167d.png)
![image](https://user-images.githubusercontent.com/18193301/158501413-504aa7b7-1c2c-446e-80ef-4224c840d67c.png)
![image](https://user-images.githubusercontent.com/18193301/158502551-1f4d2a06-8aff-4dd4-9242-9f99ed20dfbe.png)
![image](https://user-images.githubusercontent.com/18193301/158503491-58a80173-212b-49b2-8737-63d95bb40ddd.png)
![image](https://user-images.githubusercontent.com/18193301/158504842-de2b0e21-cbd3-44ab-b836-f403634e985d.png)






Side effects

- It has no effect if it is turned off by default. After turning on the limit, it may affect the maximum qps value.

### Release note 

```release-note
Add a global foreground quota limiter to control cpu and bandwidth usage, which will improve the stability of QPS.
```
